### PR TITLE
Move most of `TyCtxtAt::$name` into a generic `evaluate_query` function

### PR DESCRIPTION
**Pull request by jyn514**
Helps with https://github.com/rust-lang/rust/issues/96524. This is based on https://github.com/rust-lang/rust/pull/100943, so it looks larger than it is.

- Move `Queries::new` out of the `define_callbacks` macro. This is a very small cleanup, happy to move it to its own PR.
- Move most of `TyCtxtAt::$name` into a generic function. There were a couple different approaches I could have taken here, see https://rust-lang.zulipchat.com/#narrow/stream/241847-t-compiler.2Fwg-incr-comp/topic/making.20.60TyCtxtAt.3A.3A.24query.60.20generic.20.2396524. I chose to use a `QueryCache` generic type argument and pass in `tcx.query_caches.$name` instead of doing a much larger refactor to define `queries::$name` in rustc_middle instead of rustc_query_impl.
- Remove `OnHit` generic parameter to `try_get_cached`. It was always used with `copy` after my refactor.

r? @cjgillot

### CODE REVIEW

File: compiler/rustc_middle/src/ty/query.rs [hunk 1]
```diff
@@ -246,17 +257,10 @@ macro_rules! define_callbacks {
             #[inline(always)]
             pub fn $name(self, key: query_helper_param_ty!($($K)*)) -> query_stored::$name<'tcx>
             {
-                let key = key.into_query_param();
+                let key: $( $K )* = key.into_query_param();
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/query.rs:260]: Why do we need to specify the type here and not in the `Ensure` case?

Comment (jyn514) [reply compiler/rustc_middle/src/ty/query.rs:260]: No reason, I added it when trying to get this code to compile and forgot to remove it.

**Comment by cjgillot** [comment]
Do you have an idea where the perf regression might come from?  I'd suspect the extra coercion to `fn` ptr.  Should we re-run perf?

**Comment by jyn514** [comment]
@cjgillot hmm, we're now copying the `Cache::Stored` value for evaluate_query in QueryMode::Ensure, where we weren't before. Maybe that has a perf impact?

@bors try @rust-timer

**Comment by jyn514** [comment]
I can also split https://github.com/rust-lang/rust/pull/101178/commits/d2c53caee218296482c11a56354688a9a7e9636f to a separate PR to rule out any perf impact, I'll do that this evening.

**Comment by cjgillot** [comment]
@rust-timer build 43ab395f7c492c9ec65610e092ad26cb9879373e

**Comment by jyn514** [comment]
If this doesn't work, I think I may just close the PR; this seems like unnecessarily much work for how little duplication it removes.

File: compiler/rustc_middle/src/ty/query.rs [hunk 2]
```diff
@@ -102,14 +102,28 @@ impl<'tcx> TyCtxt<'tcx> {
...
 
-/// Helper for `TyCtxtEnsure` to avoid a closure.
-#[inline(always)]
-fn noop<T>(_: &T) {}
-
-/// Helper to ensure that queries only return `Copy` types.
-#[inline(always)]
-fn copy<T: Copy>(x: &T) -> T {
-    *x
+fn evaluate_query<'tcx, Cache>(
```
Comment (Zoxc) [inline compiler/rustc_middle/src/ty/query.rs:105]: Did you try this with `#[inline(always)]`? The callers all have that so this should regress without it.

**Pull request closed without merging**